pallets/fragments/src/lib.rs

```diff
@@ -1538,6 +1549,277 @@ pub mod pallet {
+			amount: u128,
+			currency: Option<T::AssetId>
+		) -> DispatchResult {
+			if let Some(currency) = currency {
+				<pallet_assets::Pallet<T> as fungibles::Transfer<T::AccountId>>::transfer(
+					// transfer `amount` units of `currency` from `from` to `to`
+					currency,
+					from,
+					to,
+					amount.saturated_into(),
+					true, // The debited account must stay alive at the end of the operation; an error is returned if this cannot be achieved legally.
+				)
+					.map_err(|_| Error::<T>::InsufficientBalance)?;
+			} else {
+				pallet_balances::Pallet::<T>::do_transfer(
+					// transfer `amount` units of NOVA from `from` to `to`
+					from,
+					to,
+					amount.saturated_into(),
+					ExistenceRequirement::KeepAlive,
+				)
+					.map_err(|_| Error::<T>::InsufficientBalance)?;
+			}
+
+			Ok(())
+		}
+
+		/// Whether a Fragment Instance can be transferred
+		///
+		/// * `from` - Clamor Account ID to transfer the Fragment Instance from
+		/// * `definition_hash` - Fragment Definition of the Fragment Instance
+		/// * `edition_id` - Edition ID of the Fragment Instance
+		/// * `copy_id` - Copy ID of the Fragment Instance
+		/// * `new_permissions` - New Permissions of the Fragment Instance after it is sold
+		/// * `expiration` -  Block number that the newly-copied Fragment Instance expires at. If the Fragment Instance is not copyable, this field is practically irrelevant.
+		pub fn can_transfer_instance(
+			from: &T::AccountId,
+			definition_hash: &Hash128,
+			edition_id: Unit,
+			copy_id: Unit,
+			new_permissions: Option<FragmentPerms>,
+			expiration: Option<T::BlockNumber>,
+		) -> DispatchResult {
+
+			let current_block_number = <frame_system::Pallet<T>>::block_number();
+
+			let item_data = <Fragments<T>>::get((definition_hash, edition_id, copy_id))
+				.ok_or(Error::<T>::NotFound)?;
+
+			// no go if will expire this block
+			if let Some(item_expiration) = item_data.expiring_at {
+				ensure!(current_block_number < item_expiration, Error::<T>::NotFound);
+			}
+
+			if let Some(expiration) = expiration {
+				ensure!(current_block_number < expiration, Error::<T>::ParamsNotValid);
+			}
+
+			// Only the owner of this fragment can transfer it
+			let ids =
+				<Inventory<T>>::get(from.clone(), definition_hash).ok_or(Error::<T>::NoPermission)?;
+
+			ensure!(ids.contains(&(Compact(edition_id), Compact(copy_id))), Error::<T>::NoPermission);
+
+			// first of all make sure the item can be transferred
+			ensure!(
+				(item_data.permissions & FragmentPerms::TRANSFER) == FragmentPerms::TRANSFER,
+				Error::<T>::NoPermission
+			);
+
+			if let Some(new_perms) = new_permissions {
+				// ensure we only allow more restrictive permissions
+				if (item_data.permissions & FragmentPerms::EDIT) != FragmentPerms::EDIT {
+					ensure!(
+						(new_perms & FragmentPerms::EDIT) != FragmentPerms::EDIT,
+						Error::<T>::NoPermission
+					);
+				}
+				if (item_data.permissions & FragmentPerms::COPY) != FragmentPerms::COPY {
```

Comment:
Shouldn't be `else if` in the 2nd and 3rd statements?

Reply:
the 3 if statements are:
- if `item.permissions` doesn't contain `FragmentPerms::EDIT`, then `new_perms` shouldn't either
- if `item.permissions` doesn't contain `FragmentPerms::COPY`, then `new_perms` shouldn't either
- if `item.permissions` doesn't contain `FragmentPerms::TRANSFER`, then `new_perms` shouldn't either